### DISCUSSION TITLE
SUBSTRING and SUBSTCUT instructions added

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#209] [Progress Updates]
Following @adesutherland suggestion, I added 2 new instructions commit 58d32a84817c1d4127d315db878287c1a360bea2:
```
1/ SUBSTRING target,source,offset          transfers the source string beginning at offset to the target string   
2/ SUBSTCUT  source,cut-off-position       cuts off characters beginning after cut-off-position 
                                           it's only setting the string-length, therefore it's quick

```
For both instructions: there is no padding, no length checking, if wrongly used a program exception may occur.  

These instructions may be the first step in reducing the usage of the SUBSTR function

**Comment by Peter-Jacob** [comment]
And here is the unfair contest, loop with 10,000,000 iterations:
```
time substring instruction 6.9993 
time substring function 31.9968 
```
This is the rexx:
```
do i=1 to 10000000
     assembler substring result,b,pos  /* copy string from offset pos to result */
end
say "time substring instruction "time('e')   
do i=1 to 10000000
     result=substr(b,pos)
end
say "time substring function "time('e')   
```

**Comment by Peter-Jacob** [comment]
After adding a PADSTR instruction, we can easily build a LEFT/RIGHT replacement:
```
/* fast LEFT(string,6,'0') coding */
  padnr=16
  left=6
  pad="0"
  assembler padstr result,pad,padnr
  number=number||result
  assembler substcut number,left
  say number
/* fast RIGHT(string,15,'0') coding */
  padnr=16
  right=16
  pad="+"
  string="Fred and Mary"
  assembler padstr result,pad,padnr
  string=result||string
  assembler strlen len,string
  right=len-right
  assembler substring result,string,right
  say "'"result"'"
```
Result is:
```
471100
'+++Fred and Mary'
```
Even if it is a couple of instructions, there are no loops, so it should be faster!
And could be a proposal for the CREXX precompiler hiding the instruction sequences,

**Reply by adesutherland** [reply]
Brilliant!

And as for the preprocessor - for me this should be the target for the first practical rexx level b program ... so we need some io ...

**Comment by Peter-Jacob** [comment]
you say it, I/O is the prerequisite!
